feat(dgeni): render API source block

griest024 opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/graycoreio/daffodil/blob/develop/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Fixes: #3306


## What is the new behavior?
One of the major outstanding issues with this is that the dgeni TS package doesn't correctly infer types of everything that TS can. For example, injection tokens and providers that are destructured from the factory function are untyped. Functions that do not have explicit return annotations have untyped return types even when TS correctly infers the return type.

Some of the `const`s that are function do have some of their type correctly inferred though.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

griest024 commented:
@xelaint @damienwebdev this is ready to preview and give feedback

xelaint commented:
Is it possible to make the individual properties within the source block linkable to the properties section? When a user is on a page with a long source block section (i.e. `/docs/api/cart/state/DaffCartFacade`), the individual properties take a while to scroll to.

xelaint commented on tools/dgeni/src/templates/api/base.template.html:
```diff
@@ -2,5 +2,7 @@
 <h1>{$ doc.name $}</h1>
 {% endblock %}
 <p>{$ doc.description $}</p>
+<h2>API</h2>
```
I don't think it's necessary to have this heading?

griest024 replied:
While I'm not attached to this specific heading, I think some separator between the description and the source block is necessary. When the descriptions includes some code blocks at the end (which is fairly common), I think its not immediately apparent that the source block is not part of the description. See `/docs/api/auth/state/daffAuthProvideUnauthenticatedHooks`:
![image](https://github.com/user-attachments/assets/772ed2ba-93ec-445c-9d84-fdb36b180344)

xelaint replied:
Yeah, I agree there should be a separator then. Is the `providers` code block a usage example? If so, I think we should also label it as usage example as well, or split up usage example and api into tabs like how angular does it?

griest024 replied:
yes the provider block is a usage example. I agree, we should tag it as [`@example`](https://jsdoc.app/tags-example) and render it separately (tabs sound good)

xelaint replied:
> yes the provider block is a usage example. I agree, we should tag it as [`@example`](https://jsdoc.app/tags-example) and render it separately (tabs sound good)

Sounds good. @damienwebdev likes the idea of tabs as well.

griest024 commented:
> Is it possible to make the individual properties within the source block linkable to the properties section? When a user is on a page with a long source block section (i.e. `/docs/api/cart/state/DaffCartFacade`), the individual properties take a while to scroll to.

yes this is my eventual goal. I'm not sure if it should be in this PR though since it will require changes to how the properties are laid out

xelaint commented:
Do we need to render the API blocks for design?

<img width="1081" alt="Screen Shot 2024-12-31 at 11 04 05 AM" src="https://github.com/user-attachments/assets/6047fcaa-165f-40e7-bfe0-587c19981592" />

griest024 commented:
> Do we need to render the API blocks for design?

I like them but I'll leave the final decision up to you and @damienwebdev. I have them here mostly as an example of what the source blocks will look like for components.

griest024 committed:
feat(dgeni): link symbols without link tag

The link tag will bleed into other parts of the content without being rendered. Instead, directly render symbols to anchors. This will later allow us to easily change how these symbols are rendered, e.g. as an angular router link.

griest024 committed:
feat(dgeni): extract html bracket utils

griest024 committed:
feat(dgeni): add API source block
